### ISSUE TITLE
Error while executing db2topg.pl script for DB2 sample schema

### ISSUE THREAD

**Issue by SaanSingh** [dalibo/db2topg#9]
Hello,
I am testing db2topg and hitting below error. Can please help me in this regard.

Error:
I don't understand <CREATE SUMMARY TABLE ADEFUSR AS (SELECT workdept, count(*) AS no_of_employees FROM employee GROUP BY workdept) DATA INITIALLY DEFERRED REFRESH IMMEDIATE ENABLE QUERY OPTIMIZATION MAINTAINED BY SYSTEM IN "USERSPACE1" > at ./db2topg.pl line 1092, <IN> line 450.

Thanks 
Rajni

**Comment by SaanSingh** [comment]
My Setups are like below:

I am running DB2 10.1 on Centos7 and PostgreSQL 9.5 on centos 7. Both the databases are running on same server.

I have just created sample DB2 schema and using db2topg to convert it into postgresql.

I am using below commands:
1- db2look -d sample -e -l -xd -o my_db2_sql_script
2- ./db2topg.pl -f my_db2_sql_script.sql -o my_output_dir -d sample -u db2inst2 -p xxx

After step 2 ,I am facing below issues. I tried to comment the issues encountered one by one and every time hitting new issues.

Issue1:
 ./db2topg.pl -f my_db2_sql_script.sql -o my_output_dir -d sample -u db2inst2 -p target

I don't understand <CREATE SUMMARY TABLE ADEFUSR AS (SELECT workdept, count(*) AS no_of_employees FROM employee GROUP BY workdept) DATA INITIALLY DEFERRED REFRESH IMMEDIATE ENABLE QUERY OPTIMIZATION MAINTAINED BY SYSTEM IN "USERSPACE1" > at ./db2topg.pl line 1092, <IN> line 450.

Issue2:
 ./db2topg.pl -f my_db2_sql_script.sql -o my_output_dir -d sample -u db2inst2 -p target

I don't understand   ORGANIZE BY (  in a CREATE TABLE section at ./db2topg.pl line 662, <IN> line 472.

Issue3:
./db2topg.pl -f my_db2_sql_script.sql -o my_output_dir -d sample -u db2inst2 -p target
I don't understand              GENERATE KEY USING XMLPATTERN '/product/description/detail' in an CREATE INDEX section at ./db2topg.pl line 901, <IN> line 509


Can you please help me to resolve these issues as every time I am hitting new issues after removing/commenting the previous issues. Tool is terminating after giving above errors and not doing anything.These are basic commands in DB2 and I am just testing this tool with sample schema which is very small and not having complex objects. I was expecting that this tool will move on and finally give all the errors in unsure.sql but it is terminating at very first error.

Is there any guide that I can refer to??

Thanks
Rajni

**Comment by diwathenettai** [comment]
I am getting a similar error below:

dwdev@hitdwdev01:dwdev:$ ./db2topg.pl -f my_db2_sql_script.sql -o ./my_output_dir
I don't understand              FILE '/devdata03/dwdev/dwdev/crs_cache_data.dms' 44541, in a CREATE TABLESPACE section at ./db2topg.pl line 441, <IN> line 55.
dwdev@hitdwdev01:dwdev:$ ls -lrt ./my_output_dir
total 0



The line 55 in the file that is complaining about is below:


   +33
   +34  CREATE LARGE TABLESPACE "CRS_CACHE_DATA" IN DATABASE PARTITION GROUP IBMDEFAULTGROUP PAGESIZE 32768 MANAGED BY DATABASE
   +35           USING (FILE '/devdata01/dwdev/dwdev/crs_cache_data.dms' 44541,
   +36                  FILE '/devdata02/dwdev/dwdev/crs_cache_data.dms' 44541,
   +37                  FILE '/devdata03/dwdev/dwdev/crs_cache_data.dms' 44541,
   +38                  FILE '/devdata04/dwdev/dwdev/crs_cache_data.dms' 44541,
   +39                  FILE '/devdata05/dwdev/dwdev/crs_cache_data.dms' 44541,
   +40                  FILE '/devdata06/dwdev/dwdev/crs_cache_data.dms' 44541,
   +41                  FILE '/devdata07/dwdev/dwdev/crs_cache_data.dms' 44541,
   +42                  FILE '/devdata08/dwdev/dwdev/crs_cache_data.dms' 44541,
   +43                  FILE '/devdata09/dwdev/dwdev/crs_cache_data.dms' 44541,
   +44                  FILE '/devdata10/dwdev/dwdev/crs_cache_data.dms' 44541,
   +45                  FILE '/devdata11/dwdev/dwdev/crs_cache_data.dms' 44541,
   +46                  FILE '/devdata12/dwdev/dwdev/crs_cache_data.dms' 44541)
   +47           EXTENTSIZE 32
   +48           PREFETCHSIZE 64
   +49           BUFFERPOOL BP_32K
   +50           OVERHEAD 24.100000
   +51           TRANSFERRATE 0.900000
   +52           AUTORESIZE YES
   +53           MAXSIZE NONE
   +54           FILE SYSTEM CACHING
   +55           DROPPED TABLE RECOVERY ON;
   +56

**Comment by SaanSingh** [comment]
Hello ,
My issue is still there. Can you please give some insight ??

Any help will be appreciated .

BTW diwathenettai , can you please help me to run db2topg for sample schema and let me know the work around for my problems.

Any help will be appreciated .
Thanks

**Comment by diwathenettai** [comment]
Looks like we can use the db2topg.pl for only a few object types.... I have tried removing everything else other than a few tables DDL's from the file my_db2_sql_script.sql that I require and ran the script db2topg.pl .... It seems to have generated the files for me.

**Comment by marco44** [comment]
Hi, and sorry for the delay.

Yes there are some objects that are not managed. DB2 is quite peculiar, and anyway many are simply ignored. For instance, summary tables are similar to materialized views, but specific enough that trying to migrate them will fail anyhow (the query used to generate them will have a different syntax).

For the large tablespace, there is probably an option which has not been seen until now.

Anyway, to solve your problem, please send me the sql dump to cousinmarc at gmail dot com...

**Comment by SaanSingh** [comment]
No problem Marc,
I have sent you dump.
Thanks
Rajni

**Comment by marco44** [comment]
Hi, I just pushed a new version. Please give it a try

**Comment by SaanSingh** [comment]
Hello,
Thanks for the update.
I tried it,now it is not terminating at first error and creating all the .sql files as expected but it is just including table in before.sql file and everything else is sent to unsure.sql.

Does it mean,it can only migrate tables and nothing else??

Can we have something of the sort where db2tppg can do changes like data types and other structural changes as per postgresql and cover most of the objects??


Please advice,as this is just very small and sample schema and If I am planning to use db2topostgre with the production schema then the manual task will be too much.


Thanks
Rajni

**Comment by marco44** [comment]
There are also most indexes in after.sql.

It migrates tables, indexes. It tries to migrate constraints, but this is limited as both engines are very different.

> Can we have something of the sort where db2tppg can do changes like data types and other structural changes as per postgresql and cover most of the objects??

Yes, but you'll have to write it yourself. And you'll have a lot of pain in the process, I think.

> Please advice,as this is just very small and sample schema and If I am planning to use db2topostgre with the production schema then the manual task will be too much.

It tries to migrate tables, indexes, domains, unique, PK and FK constraints, identities, and more likely to fail: check constraints,views. Functions and triggers will have to be re-written.

**Comment by SaanSingh** [comment]
Thank you so much for the help.
We can close this issue.

Thanks
